packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,436 @@
+  readonly tradeType: RouterTradeType = RouterTradeType.PancakeSwapTrade
+
+  readonly type: TradeType
+
+  constructor(public trade: AnyTradeType, public options: PancakeSwapOptions) {
+    this.type = (this.trade as SmartRouterTrade<TradeType>).tradeType
+  }
+
+  encode(planner: RoutePlanner, _config: TradeConfig): void {
+    let payerIsUser = true
+    const trades = !Array.isArray(this.trade) ? [this.trade] : this.trade
+    const numberOfTrades = trades.reduce((numOfTrades, trade) => numOfTrades + trade.routes.length, 0)
+
+    const sampleTrade = trades[0]
+    // If the input currency is the native currency, we need to wrap it with the router as the recipient
+    if (sampleTrade.inputAmount.currency.isNative) {
+      // TODO: optimize if only one v2 pool we can directly send this to the pool
+      planner.addCommand(CommandType.WRAP_ETH, [
+        ROUTER_AS_RECIPIENT,
+        BigInt(
+          SmartRouter.maximumAmountIn(
+            sampleTrade,
+            this.options.slippageTolerance,
+            sampleTrade.inputAmount,
+          ).quotient.toString(),
+        ),
+      ])
+      // since WETH is now owned by the router, the router pays for inputs
+      payerIsUser = false
+    }
+    // The overall recipient at the end of the trade, SENDER_AS_RECIPIENT uses the msg.sender
+    this.options.recipient = this.options.recipient ?? SENDER_AS_RECIPIENT
+
+    const inputIsNative = sampleTrade.inputAmount.currency.isNative
+    const outputIsNative = sampleTrade.outputAmount.currency.isNative
+    const performAggregatedSlippageCheck = sampleTrade.tradeType === TradeType.EXACT_INPUT && numberOfTrades > 2
+    const routerMustCustody = outputIsNative || !!this.options.fee || performAggregatedSlippageCheck
+
+    for (const trade of trades) {
+      if (trade.routes.length === 1 && trade.routes[0].type === RouteType.V2) {
+        addV2Swap(
+          planner,
+          trade as SmartRouterTrade<TradeType>,
+          this.options,
+          routerMustCustody,
+          payerIsUser,
+          performAggregatedSlippageCheck,
+        )
+      } else if (trade.routes.every((r) => r.type === RouteType.V3)) {
+        addV3Swap(
+          planner,
+          trade as SmartRouterTrade<TradeType>,
+          this.options,
+          routerMustCustody,
+          payerIsUser,
+          performAggregatedSlippageCheck,
+        )
+      } else {
+        addMixedSwap(
+          planner,
+          trade as SmartRouterTrade<TradeType>,
+          this.options,
+          payerIsUser,
+          routerMustCustody,
+          performAggregatedSlippageCheck,
+        )
+      }
+    }
+    // const ZERO_OUT: CurrencyAmount<Currency> = CurrencyAmount.fromRawAmount(sampleTrade.outputAmount.currency, 0)
+    // const minAmountOut: CurrencyAmount<Currency> = trades.reduce(
+    //   (sum, trade) => sum.add(minimumAmountOut(trade, options.slippageTolerance)),
+    //   ZERO_OUT,
+    // )
```

Comment:
why are we changing this implementation?

Reply:
im following uniswap protocol as amuch as i can before changing any code. the rule i was following was if i can following their base code first, then do so. if not write custom code (in whih case i would stick as close as i can to smart order router)